In [487]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings("ignore")

df = pd.read_csv('../data/mental-heath-in-tech-2016_20161114.csv')
df

Are you self-employed?  \
0                          0   
1                          0   
2                          0   
3                          1   
4                          0   
5                          0   
6                          0   
7                          0   
8                          0   
9                          1   
10                         0   
11                         0   
12                         0   
13                         0   
14                         0   
15                         0   
16                         0   
17                         0   
18                         1   
19                         0   
20                         0   
21                         0   
22                         0   
23                         0   
24                         1   
25                         0   
26                         0   
27                         0   
28                         0   
29                         0   
...                      ...   
1403                       0   
1404                       1   
1405                       0   
1406                       0   
1407                       0   
1408                       1   
1409                       0   
1410                       0   
1411                       0   
1412                       0   
1413                       0   
1414                       0   
1415                       0   
1416                       0   
1417                       0   
1418                       0   
1419                       0   
1420                       1   
1421                       0   
1422                       0   
1423                       0   
1424                       0   
1425                       0   
1426                       0   
1427                       0   
1428                       1   
1429                       1   
1430                       0   
1431                       0   
1432                       0   

     How many employees does your company or organization have?  \
0                                                26-100           
1                                                  6-25           
2                                                  6-25           
3                                                   NaN           
4                                                  6-25           
5                                        More than 1000           
6                                                26-100           
7                                        More than 1000           
8                                                26-100           
9                                                   NaN           
10                                               26-100           
11                                              100-500           
12                                              100-500           
13                                              100-500           
14                                              100-500           
15                                              100-500           
16                                               26-100           
17                                             500-1000           
18                                                  NaN           
19                                              100-500           
20                                                 6-25           
21                                       More than 1000           
22                                               26-100           
23                                                  1-5           
24                                                  NaN           
25                                              100-500           
26                                               26-100           
27                                              100-500           
28                                             500-1000           
29         

## These column headers are gross, and we're only going to look at a few of them...

In [488]:
data = np.array([df['What is your age?'],
        df['What is your gender?'],
        df['What country do you live in?'],
        df['What country do you work in?'],
        df['How many employees does your company or organization have?'],
        df['Has your employer ever formally discussed mental health (for example, as part of a wellness campaign or other official communication)?'],
        df['Do you have a family history of mental illness?'],
        df['Have you had a mental health disorder in the past?'],
        df['Do you currently have a mental health disorder?'],
        df['Have you ever sought treatment for a mental health issue from a mental health professional?'],
        df['Do you work remotely?']
       ])

columns= ['age',
         'gender',
         'residence_location',
         'work_location',
         'num_employees',
         'employer_awareness',
         'family_history',
         'past_history',
         'current_state',
         'treatment',
         'remote']


In [510]:
df = pd.DataFrame(data=data)
df = df.T # transpose because lists are horizontal
df.columns = columns
df

age                 gender        residence_location  \
0     39                   Male            United Kingdom   
1     29                   male  United States of America   
2     38                  Male             United Kingdom   
3     43                   male            United Kingdom   
4     43                 Female  United States of America   
5     42                   Male            United Kingdom   
6     30                      M  United States of America   
7     37                 female  United States of America   
8     44                 Female  United States of America   
9     30                   Male  United States of America   
10    28                   Male  United States of America   
11    34                   Male  United States of America   
12    30                 Female  United States of America   
13    44                   Male                    Canada   
14    35                   Male  United States of America   
15    52                      m  United States of America   
16    32  I identify as female.                    Canada   
17    34                   Male            United Kingdom   
18    25                female   United States of America   
19    31                      M  United States of America   
20    26                   Male  United States of America   
21    35                   male  United States of America   
22    25                   male                   Germany   
23    33                   Male                    Canada   
24    38                   Male  United States of America   
25    42                   male                   Germany   
26    27                   Male               Netherlands   
27    35                   male                   Germany   
28    34                 female            United Kingdom   
29    32               Bigender            United Kingdom   
...   ..                    ...                       ...   
1403  27                   Male                     India   
1404  29                   Male                    Canada   
1405  28                   Male                    Poland   
1406  26      Transgender woman                   Ireland   
1407  29                   male               Afghanistan   
1408  34                      m                   Germany   
1409  36                   Male            United Kingdom   
1410  38                   Male                   Belgium   
1411  22                 Female  United States of America   
1412  24                   male  United States of America   
1413  24                 female  United States of America   
1414  36                 female  United States of America   
1415  24                 Female  United States of America   
1416  24                   Male  United States of America   
1417  31                   Male               Netherlands   
1418  39                      M            United Kingdom   
1419  41                   male  United States of America   
1420  29                   Male            United Kingdom   
1421  37                   male                   Estonia   
1422  30                   male                    Canada   
1423  24                   Male                    Canada   
1424  22                 female                    Canada   
1425  24                   Male                    Canada   
1426  26                 Female                    Canada   
1427  38                 female  United States of America   
1428  34                Female   United States of America   
1429  56                   MALE  United States of America   
1430  52                   Male  United States of America   
1431  30                 Female  United States of America   
1432  25             non-binary                    Canada   

                 work_location   num_employees employer_awareness  \
0               United Kingdom          26-100                 No   
1     United States of America            6-25                Yes   
2              

In [511]:
df.info()
df.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1433 entries, 0 to 1432
Data columns (total 11 columns):
age                   1433 non-null object
gender                1430 non-null object
residence_location    1433 non-null object
work_location         1433 non-null object
num_employees         1146 non-null object
employer_awareness    1146 non-null object
family_history        1433 non-null object
past_history          1433 non-null object
current_state         1433 non-null object
treatment             1433 non-null object
remote                1433 non-null object
dtypes: object(11)
memory usage: 123.2+ KB


age gender        residence_location             work_location  \
count   1433   1430                      1433                      1433   
unique    53     70                        53                        53   
top       30   Male  United States of America  United States of America   
freq      94    610                       840                       851   

       num_employees employer_awareness family_history past_history  \
count           1146               1146           1433         1433   
unique             6                  3              3            3   
top           26-100                 No            Yes          Yes   
freq             292                813            670          736   

       current_state  treatment     remote  
count           1433       1433       1433  
unique             3          2          3  
top              Yes          1  Sometimes  
freq             575        839        757

# Clean things up a bit

### Age

In [514]:
# lets check for any wonky entries in age
df[df.age.apply(lambda x: type(x)!=np.int64)]

Empty DataFrame
Columns: [age, gender, residence_location, work_location, num_employees, employer_awareness, family_history, past_history, current_state, treatment, remote]
Index: []

good all numbers

In [517]:
# casting as an int
df.age = df.age.astype(np.int64)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1433 entries, 0 to 1432
Data columns (total 11 columns):
age                   1433 non-null int64
gender                1430 non-null object
residence_location    1433 non-null object
work_location         1433 non-null object
num_employees         1146 non-null object
employer_awareness    1146 non-null object
family_history        1433 non-null object
past_history          1433 non-null object
current_state         1433 non-null object
treatment             1433 non-null object
remote                1433 non-null object
dtypes: int64(1), object(10)
memory usage: 123.2+ KB


In [518]:
df.age.describe()

count    1433.000000
mean       34.286113
std        11.290931
min         3.000000
25%        28.000000
50%        33.000000
75%        39.000000
max       323.000000
Name: age, dtype: float64

Looks like there a few weirdos.  I doubt a 3 yo participated in this survey, and I'm pretty sure no one is actually 323.  Lets restrict the ages from 18 - 65.

In [519]:
df.drop(df[(df.age<16) | (df.age > 80)].index, inplace=True)
df.describe()

age
count  1429.000000
mean     34.074178
std       8.096068
min      17.000000
25%      28.000000
50%      33.000000
75%      39.000000
max      74.000000

Much better.

### Gender
lets look at the gender field

In [520]:
df.gender.describe()

count     1426
unique      69
top       Male
freq       608
Name: gender, dtype: object

wow 70 unique types.  Okay, I support that.  Lets take a look

In [521]:
set(df.gender)

{nan,
 'Unicorn',
 'Androgynous',
 'Agender',
 'Man',
 'I identify as female.',
 'genderqueer woman',
 'mtf',
 'Cis female ',
 'Malr',
 'female',
 'M',
 'Male',
 'cisdude',
 'cis man',
 'male',
 'human',
 'genderqueer',
 'MALE',
 'Genderfluid (born female)',
 'Woman',
 'Dude',
 'non-binary',
 "I'm a man why didn't you make this a drop down question. You should of asked sex? And I would of answered yes please. Seriously how much text can this take? ",
 'mail',
 'Queer',
 ' Female',
 'Female ',
 'female/woman',
 'female ',
 'Cisgender Female',
 'm',
 'Male.',
 'Cis-woman',
 'Genderqueer',
 'Male/genderqueer',
 'Transgender woman',
 'Male ',
 'Male (cis)',
 'Other/Transfeminine',
 'Sex is male',
 'woman',
 'fem',
 'none of your business',
 'Male (trans, FtM)',
 'Genderflux demi-girl',
 'Female or Multi-Gender Femme',
 'Female assigned at birth ',
 'Transitioned, M2F',
 'Cis Male',
 'AFAB',
 'Nonbinary',
 'Fluid',
 'Genderfluid',
 'Enby',
 'Cis male',
 'Female (props for making this a free

Wow.  Okay that's great. Lets order it, hopefully something useful will appear.

In [522]:
df.groupby('gender').count().sort_values(by='age', ascending=False)

age  residence_location  \
gender                                                                        
Male                                                608                 608   
male                                                248                 248   
Female                                              153                 153   
female                                               95                  95   
M                                                    86                  86   
m                                                    79                  79   
F                                                    38                  38   
f                                                    23                  23   
Male                                                 11                  11   
Female                                                9                   9   
woman                                                 4                   4   
non-binary                                            4                   4   
female                                                3                   3   
man                                                   3                   3   
Woman                                                 3                   3   
Male (cis)                                            2                   2   
Man                                                   2                   2   
Nonbinary                                             2                   2   
male                                                  2                   2   
Agender                                               2                   2   
Transitioned, M2F                                     1                   1   
Unicorn                                               1                   1   
cis man                                               1                   1   
cisdude                                               1                   1   
cis male                                              1                   1   
fm                                                    1                   1   
fem                                                   1                   1   
female-bodied; no feelings about gender               1                   1   
female/woman                                          1                   1   
Sex is male                                           1                   1   
...                                                 ...                 ...   
 Female                                               1                   1   
Queer                                                 1                   1   
Fluid                                                 1                   1   
Androgynous                                           1                   1   
Bigender                                              1                   1   
Cis Male                                              1                   1   
Cis female                                            1                   1   
Cis male                                              1                   1   
Cis-woman                                             1                   1   
Cisgender Female                                      1                   1   
Dude                                                  1                   1   
Enby                                                  1                   1   
Female (props for making this a freeform field,...    1                   1   
Female assigned at birth                              1                   1   
Female or Multi-Gender Femme                          1                   1   
Genderfluid                                           1                   1   
Other/Transfeminine                                   1                   1   
Genderfluid (born female)                             1                   1   
Genderflux demi-girl   

okay lets take [Male,male,M,m] [Female,female,F,f] to male female, and drop everything else... sorry I support you, but this will make things WAY simpler

In [523]:
# replace varieties of male and female to 0 and 1
df.gender.replace(to_replace=['Male', 'M', 'male', 'm'], value=0, inplace=True)
df.gender.replace(to_replace=['Female', 'F', 'female', 'f'], value=1, inplace=True)

# remove all other observations (sorry ya'll)
df = df[df.gender.isin([0,1])]
df.gender = df.gender.astype(bool)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1330 entries, 0 to 1431
Data columns (total 11 columns):
age                   1330 non-null int64
gender                1330 non-null bool
residence_location    1330 non-null object
work_location         1330 non-null object
num_employees         1067 non-null object
employer_awareness    1067 non-null object
family_history        1330 non-null object
past_history          1330 non-null object
current_state         1330 non-null object
treatment             1330 non-null object
remote                1330 non-null object
dtypes: bool(1), int64(1), object(9)
memory usage: 115.6+ KB


### number of employees

looks like there some missing data for number of employees

It would be sort of difficult to impute that from the other attributes, so we'll just drop those instances

In [509]:
df.dropna(subset=['num_employees'], inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1067 entries, 0 to 1431
Data columns (total 11 columns):
age                   1067 non-null int64
gender                1067 non-null bool
residence_location    1067 non-null object
work_location         1067 non-null object
num_employees         1067 non-null object
employer_awareness    1067 non-null object
family_history        1067 non-null object
past_history          1067 non-null object
current_state         1067 non-null object
treatment             1067 non-null object
remote                1067 non-null object
dtypes: bool(1), int64(1), object(9)
memory usage: 92.7+ KB


What do you know?  That took care of the null 'employer_awareness' field too.

In [525]:
df

age  gender        residence_location             work_location  \
0      39   False            United Kingdom            United Kingdom   
1      29   False  United States of America  United States of America   
3      43   False            United Kingdom            United Kingdom   
4      43    True  United States of America  United States of America   
5      42   False            United Kingdom            United Kingdom   
6      30   False  United States of America  United States of America   
7      37    True  United States of America  United States of America   
8      44    True  United States of America  United States of America   
9      30   False  United States of America  United States of America   
10     28   False  United States of America  United States of America   
11     34   False  United States of America  United States of America   
12     30    True  United States of America  United States of America   
13     44   False                    Canada                    Canada   
14     35   False  United States of America  United States of America   
15     52   False  United States of America  United States of America   
17     34   False            United Kingdom            United Kingdom   
19     31   False  United States of America  United States of America   
20     26   False  United States of America  United States of America   
21     35   False  United States of America  United States of America   
22     25   False                   Germany                   Germany   
23     33   False                    Canada                    Canada   
24     38   False  United States of America  United States of America   
25     42   False                   Germany                   Germany   
26     27   False               Netherlands               Netherlands   
27     35   False                   Germany                   Germany   
28     34    True            United Kingdom            United Kingdom   
31     30   False  United States of America  United States of America   
32     31   False  United States of America  United States of America   
34     36   False            United Kingdom            United Kingdom   
35     33   False                   Germany                   Germany   
...   ...     ...                       ...                       ...   
1399   28   False  United States of America  United States of America   
1400   28    True  United States of America  United States of America   
1401   37   False  United States of America  United States of America   
1402   30   False  United States of America  United States of America   
1403   27   False                     India                     India   
1404   29   False                    Canada                    Canada   
1405   28   False                    Poland                    Poland   
1407   29   False               Afghanistan  United States of America   
1408   34   False                   Germany                   Germany   
1409   36   False            United Kingdom            United Kingdom   
1410   38   False                   Belgium                   Belgium   
1411   22    True  United States of America  United States of America   
1412   24   False  United States of America  United States of America   
1413   24    True  United States of America  United States of America   
1414   36    True  United States of America  United States of America   
1415   24    True  United States of America  United States of America   
1416   24   False  United States of America  United States of America   
1417   31   False               Netherlands               Netherlands   
1418   39   False            United Kingdom            United Kingdom   
1419   41   False  United States of America  United States of America   
1420   29   False            United Kingdom            United Kingdom   
1421   37   False                   Estonia                   Estonia   
1422   30   False                    Canada  United States of Am

**TODO** should we change from the string representation of a range to something else

In [387]:
set(df.num_employees)

{'1-5', '100-500', '26-100', '500-1000', '6-25', 'More than 1000'}

In [388]:
df.groupby('num_employees').count()

age  gender  residence_location  work_location  \
num_employees                                                    
1-5              55      55                  55             55   
100-500         234     234                 234            234   
26-100          271     271                 271            271   
500-1000         73      73                  73             73   
6-25            195     195                 195            195   
More than 1000  239     239                 239            239   

                employer_awareness  family_history  past_history  \
num_employees                                                      
1-5                             55              55            55   
100-500                        234             234           234   
26-100                         271             271           271   
500-1000                        73              73            73   
6-25                           195             195           195   
More than 1000                 239             239           239   

                current_state  treatment  remote  
num_employees                                     
1-5                        55         55      55  
100-500                   234        234     234  
26-100                    271        271     271  
500-1000                   73         73      73  
6-25                      195        195     195  
More than 1000            239        239     239

### recast treatment

In [389]:
df.treatment = df.treatment.astype(bool)


## Let's Check it out

In [390]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1067 entries, 4 to 1431
Data columns (total 11 columns):
age                   1067 non-null int64
gender                1067 non-null bool
residence_location    1067 non-null object
work_location         1067 non-null object
num_employees         1067 non-null object
employer_awareness    1067 non-null object
family_history        1067 non-null object
past_history          1067 non-null object
current_state         1067 non-null object
treatment             1067 non-null bool
remote                1067 non-null object
dtypes: bool(2), int64(1), object(8)
memory usage: 85.4+ KB


In [391]:
df

age  gender        residence_location             work_location  \
4      43    True  United States of America  United States of America   
5      42   False            United Kingdom            United Kingdom   
6      30   False  United States of America  United States of America   
7      37    True  United States of America  United States of America   
8      44    True  United States of America  United States of America   
10     28   False  United States of America  United States of America   
11     34   False  United States of America  United States of America   
12     30    True  United States of America  United States of America   
13     44   False                    Canada                    Canada   
14     35   False  United States of America  United States of America   
15     52   False  United States of America  United States of America   
17     34   False            United Kingdom            United Kingdom   
19     31   False  United States of America  United States of America   
20     26   False  United States of America  United States of America   
21     35   False  United States of America  United States of America   
22     25   False                   Germany                   Germany   
23     33   False                    Canada                    Canada   
25     42   False                   Germany                   Germany   
26     27   False               Netherlands               Netherlands   
27     35   False                   Germany                   Germany   
28     34    True            United Kingdom            United Kingdom   
31     30   False  United States of America  United States of America   
32     31   False  United States of America  United States of America   
34     36   False            United Kingdom            United Kingdom   
35     33   False                   Germany                   Germany   
36     44   False  United States of America  United States of America   
37     29   False            United Kingdom            United Kingdom   
38     29   False            United Kingdom            United Kingdom   
39     44   False                    Canada                    Canada   
41     40   False  United States of America  United States of America   
...   ...     ...                       ...                       ...   
1394   33   False            United Kingdom            United Kingdom   
1397   28   False  United States of America  United States of America   
1398   30   False  United States of America  United States of America   
1399   28   False  United States of America  United States of America   
1400   28    True  United States of America  United States of America   
1401   37   False  United States of America  United States of America   
1402   30   False  United States of America  United States of America   
1403   27   False                     India                     India   
1405   28   False                    Poland                    Poland   
1407   29   False               Afghanistan  United States of America   
1409   36   False            United Kingdom            United Kingdom   
1410   38   False                   Belgium                   Belgium   
1411   22    True  United States of America  United States of America   
1412   24   False  United States of America  United States of America   
1413   24    True  United States of America  United States of America   
1414   36    True  United States of America  United States of America   
1415   24    True  United States of America  United States of America   
1416   24   False  United States of America  United States of America   
1417   31   False               Netherlands               Netherlands   
1418   39   False            United Kingdom            United Kingdom   
1419   41   False  United States of America  United States of America   
1421   37   False                   Estonia                   Estonia   
1422   30   False                    Canada  United States of Am

# NICE!